### PR TITLE
Add check to enforce that all tokens within a group must have same Concrete Class and governing ContractClass 

### PR DESCRIPTION
**Pull request by roastario**

### CODE REVIEW

File: contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/AbstractTokenContract.kt [hunk 1]
```diff
@@ -27,6 +23,8 @@ import net.corda.core.transactions.LedgerTransaction.InOutGroup
  */
 abstract class AbstractTokenContract<T : TokenType, U : AbstractToken<T>> : Contract {
 
+    abstract val accepts: Class<U>
```
Comment (roger3cev) [inline contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/AbstractTokenContract.kt]: We can add kdocs here. I'll do it in a separate PR though.

File: contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/AbstractTokenContract.kt [hunk 2]
```diff
@@ -116,24 +111,39 @@ abstract class AbstractTokenContract<T : TokenType, U : AbstractToken<T>> : Cont
...
 
     private fun verifyTypeJarPresentInTransaction(jar: SecureHash, attachments: List<Attachment>) {
-        require(jar in attachments.map(Attachment::id)) {
-            "Expected to find type jar: $jar in transaction attachment list, but did not"
+        require(jar in attachments.map { it.id }) { "Expected to find type jar: $jar in transaction attachment list, but did not" }
+    }
+
+    private fun verifyAllTokensUseSameTypeJar(inputs: List<AbstractToken<*>>, outputs: List<AbstractToken<*>>): SecureHash? {
+        val jarHashes = (inputs + outputs).map { it.tokenTypeJarHash }.toSet()
+        require(jarHashes.size == 1) { "There must only be one Jar (Hash) providing TokenType: ${outputs.first().tokenType.tokenIdentifier}" }
```
Comment (roger3cev) [inline contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/AbstractTokenContract.kt]: Do we need to deal with token pointers here?

File: contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/AbstractTokenContract.kt [hunk 3]
```diff
@@ -73,15 +71,12 @@ abstract class AbstractTokenContract<T : TokenType, U : AbstractToken<T>> : Cont
...
-    /** Necessary for providing a grouping function, which is usually the [IssuedTokenType]. */
-    abstract fun groupStates(tx: LedgerTransaction): List<InOutGroup<U, IssuedTokenType<T>>>
-
     final override fun verify(tx: LedgerTransaction) {
         // Group token amounts by token type. We need to do this because tokens of different types need to be
         // verified separately. This works for the same token type with different issuers, or different token types
         // altogether. The grouping function returns a list containing groups of input and output states grouped by
         // token type. The type is specified explicitly to aid understanding.
-        val groups: List<InOutGroup<U, IssuedTokenType<T>>> = groupStates(tx)
+        val groups = groupStates(tx) { TokenInfo(it.data.javaClass, it.data.issuedTokenType, it.contract) }
```
Comment (roger3cev) [inline contracts/src/main/kotlin/com/r3/corda/lib/tokens/contracts/AbstractTokenContract.kt]: should we add some tests to prove we fixed the bug?

**Comment by tudor-malene** [comment]
I'm not sure I totally understand the reasoning around this.

I was expecting a simple check like this per group:
`(inputs+outputs).map(::contractClassName).toSet().single()`
`(inputs+outputs).map{data::class.java.name}.toSet().single()`

**Comment by roastario** [comment]
> I'm not sure I totally understand the reasoning around this.
> 
> I was expecting a simple check like this per group:
> `(inputs+outputs).map(::contractClassName).toSet().single()`
> `(inputs+outputs).map{data::class.java.name}.toSet().single()`

The problem with this approach is that it would prevent DvP of `FungibleToken(GBP)` -> `SanctionsCompliant(GBP)  `

What the PR does is group all by (TokenClass, TokenType, ContractClass), and the existing invariants (input.amount = output.amount, etc etc) will naturally enforce the suggested checks without having to explicitly performing them

**Comment by tudor-malene** [comment]
> The problem with this approach is that it would prevent DvP of `FungibleToken(GBP)` -> `SanctionsCompliant(GBP) `

You're saying that you can have a transaction with just 1 input state of type `FungibleToken(GBP)` and just one output of type: `SanctionsCompliant(GBP) ` ?

Or is the transaction actually:

```
input: FungibleToken , SanctionsCompliant
output: FungibleToken , SanctionsCompliant 
```

**Comment by roastario** [comment]
> > The problem with this approach is that it would prevent DvP of `FungibleToken(GBP)` -> `SanctionsCompliant(GBP) `
> 
> You're saying that you can have a transaction with just 1 input state of type `FungibleToken(GBP)` and just one output of type: `SanctionsCompliant(GBP) ` ?
> 
> Or is the transaction actually:
> 
> ```
> input: FungibleToken , SanctionsCompliant
> output: FungibleToken , SanctionsCompliant 
> ```
The second type, as in: https://github.com/corda/token-sdk/pull/95/files#diff-906e2ba78f84a6792af0a26c2edc3e06R245

**Comment by tudor-malene** [comment]
> > > The problem with this approach is that it would prevent DvP of `FungibleToken(GBP)` -> `SanctionsCompliant(GBP) `
> > 
> > 
> > You're saying that you can have a transaction with just 1 input state of type `FungibleToken(GBP)` and just one output of type: `SanctionsCompliant(GBP) ` ?
> > Or is the transaction actually:
> > ```
> > input: FungibleToken , SanctionsCompliant
> > output: FungibleToken , SanctionsCompliant 
> > ```
> 
> The second type, as in: https://github.com/corda/token-sdk/pull/95/files#diff-906e2ba78f84a6792af0a26c2edc3e06R245

Gotcha.
This scenario would not have worked correctly even if nobody would have tried to hack it.

Grouping by the extended `TokenInfo` and then checking that:
```
(inputs+outputs).map{data::class.java.name}.toSet().single()
```
should be fine.

What is the reason for the `accepts` thing?

**Comment by roastario** [comment]
> > > > The problem with this approach is that it would prevent DvP of `FungibleToken(GBP)` -> `SanctionsCompliant(GBP) `
> > > 
> > > 
> > > You're saying that you can have a transaction with just 1 input state of type `FungibleToken(GBP)` and just one output of type: `SanctionsCompliant(GBP) ` ?
> > > Or is the transaction actually:
> > > ```
> > > input: FungibleToken , SanctionsCompliant
> > > output: FungibleToken , SanctionsCompliant 
> > > ```
> > 
> > 
> > The second type, as in: https://github.com/corda/token-sdk/pull/95/files#diff-906e2ba78f84a6792af0a26c2edc3e06R245
> 
> Gotcha.
> This scenario would not have worked correctly even if nobody would have tried to hack it.
> 
> Grouping by the extended `TokenInfo` and then checking that:
> 
> ```
> (inputs+outputs).map{data::class.java.name}.toSet().single()
> ```
> 
> should be fine.
> 
> What is the reason for the `accepts` thing?

`accepts` is required, because for each TokenType group there can be both Fungible and NonFungible tokens (IE FungibleToken(GBP) and House(GBP)) and we have to make sure that only the ones which each Contract accepts are passed to it for checking

**Comment by tudor-malene** [comment]
> `accepts` is required, because for each TokenType group there can be both Fungible and NonFungible tokens (IE FungibleToken(GBP) and House(GBP)) and we have to make sure that only the ones which each Contract accepts are passed to it for checking

Sorry, still don't get it.
The new `TokenInfo` grouping logic already groups by the state type class and the `ContractClassName`.
How can you have both fungible and non-fungible in one group?

**Pull request merged**